firmware/esp32_ttgo/middleware/Console.cpp

```diff
@@ -0,0 +1,108 @@
+#include "Console.h"
+
+#include <Layer1.h>
+#include <LoRaLayer2.h>
+
+#include <vector>
+
+#define START_MESSAGE "Type '/join NICKNAME' to join the chat, or '/help' for more commands."
+
+void Console::setup()
+{
+    DisasterMiddleware::setup();
+    if (history)
+    {
+        history->replay(this);
+    }
+    client->receive(String(START_MESSAGE) + "\n");
+}
+
+void Console::processLine(String message)
+{
+    if (message.charAt(0) == '/')
+    {
+        std::vector<String> args;
+
+        int start = 0, end;
+        while (start < message.length())
+        {
+            int index = message.indexOf(' ', start);
+            end = index == -1 ? message.length() : index;
+            args.push_back(message.substring(start, end));
+            start = end + 1;
+        }
+
+        String command = args[0].substring(1);
+        if (command == "help")
+        {
+            client->receive("Commands: /help /join /nick /raw /restart\n");
+        }
+        else if (command == "raw")
+        {
+            disconnect(client);
+            server->disconnect(this);
+            server->connect(client);
+        }
+        else if ((command == "join" || command == "nick") && args.size() > 1)
+        {
+            String msg = username.length() > 0 ? String("~ ") + username + " is now known as " + args[1] : String("~ ") + args[1] + " joined the channel";
+            username = args[1];
+            server->transmit(this, String("c|") + msg);
+            client->receive(msg);
+        }
+        else if (command == "restart")
+        {
+            ESP.restart();
+        }
+        else if (command == "lora")
+        {
+            // TODO: print to client instead of serial
+            Serial.printf("Address: ");
+            LL2.printAddress(Layer1.localAddress());
+            LL2.printRoutingTable();
+        }
+        else
+        {
+            client->receive(String("Unknown command '") + command + "'\n");
+        }
+    }
+    else if (username.length() > 0)
+    {
+
+        Serial.printf("4:[%d][%s]\n", message.length(), message.c_str());
+        server->transmit(this, String("c|") + "<" + username + "> " + message);
+        client->receive(String("<") + username + ">" + " " + message + "\n");
+    }
+    else
+    {
+        client->receive(String(START_MESSAGE) + "\n");
+    }
+}
+
+void Console::transmit(DisasterClient *client, String message)
+{
+    Serial.printf("2:[%d][%s]\n", message.length(), message.c_str());
+    client->receive(message);
+
+    buffer.concat(message);
+
+    int index;
+    while ((index = buffer.indexOf('\n')) >= 0)
+    {
+        // split the string on \n or \r\n
+        String line = buffer.charAt(index - 1) == '\r'
+                          ? buffer.substring(0, index - 1)
+                          : buffer.substring(0, index);
+        Serial.printf("3:[%d][%s]\n", line.length(), line.c_str());
```

Comment:
This makes the console print out very ugly. I'm wondering  if we can bring back some of the "prettified" print out that I had written for my console, https://github.com/sudomesh/disaster-radio/blob/master/firmware/esp32_ttgo/console.cpp#L154
I also don't love the idea of using String, i does make it much easier and the console input, theoretically, is always going to be string, so I suppose it makes sense.

Reply:
That's a debugging line I accidentally left in, it can be removed.